### PR TITLE
Add tracing context to messages

### PR DESCRIPTION
**Pull request by alicefuzier**

### CODE REVIEW

File: messaging/src/main/scala/uk/ac/wellcome/messaging/sqsworker/alpakka/AlpakkaSQSWorker.scala [hunk 1]
```diff
@@ -57,15 +56,15 @@ object AlpakkaSQSWorker {
...
+    mp: MonitoringProcessor[Work, MonitoringContext],
     sc: AmazonSQSAsync,
     as: ActorSystem,
     wd: Decoder[Work]) =
     new AlpakkaSQSWorker[
       Work,
       MonitoringContext,
       Summary,
-      MonitoringProcessor[SQSMessage, MonitoringContext]](
+      MonitoringProcessor[Work, MonitoringContext]](
```
Comment (kenoir) [inline messaging/src/main/scala/uk/ac/wellcome/messaging/sqsworker/alpakka/AlpakkaSQSWorker.scala]: 💥

File: messaging/src/main/scala/uk/ac/wellcome/messaging/worker/steps/MessageTransform.scala [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import scala.util.Try
+
+trait MessageTransform[Message, Work, MonitoringContext] {
+
+  type Transformed =
+    (Either[Throwable, Work], Either[Throwable, Option[MonitoringContext]])
+
+  protected val transform: Message => Transformed
+
+  final def callTransform(message: Message): Transformed = {
```
Comment (kenoir) [inline messaging/src/main/scala/uk/ac/wellcome/messaging/worker/steps/MessageTransform.scala:12]: Explicit transform step is nice.

**Review by kenoir — APPROVED** [review]

File: messaging/src/main/scala/uk/ac/wellcome/messaging/sqsworker/alpakka/AlpakkaSQSWorker.scala [hunk 3]
```diff
@@ -14,12 +14,11 @@ import uk.ac.wellcome.messaging.worker.steps.MonitoringProcessor
...
-                       MonitoringContext,
-                       Summary,
-                       MonitoringProcessorImpl <: MonitoringProcessor[
-                         SQSMessage,
-                         MonitoringContext]](
+class AlpakkaSQSWorker[
+  Work,
+  MonitoringContext,
+  Summary,
+  MonitoringProcessorImpl <: MonitoringProcessor[Work, MonitoringContext]](
```
Comment (warrd) [inline messaging/src/main/scala/uk/ac/wellcome/messaging/sqsworker/alpakka/AlpakkaSQSWorker.scala]: I think have this parameterised is unnecessary, and is slightly easier to read with one less type parameter. This diff compiles without it:

```diff
diff --git a/messaging/src/main/scala/uk/ac/wellcome/messaging/sqsworker/alpakka/AlpakkaSQSWorker.scala b/messaging/src/main/scala/uk/ac/wellcome/messaging/sqsworker/alpakka/AlpakkaSQSWorker.scala
index 7a06fba..99ac2a4 100644
--- a/messaging/src/main/scala/uk/ac/wellcome/messaging/sqsworker/alpakka/AlpakkaSQSWorker.scala
+++ b/messaging/src/main/scala/uk/ac/wellcome/messaging/sqsworker/alpakka/AlpakkaSQSWorker.scala
@@ -14,16 +14,12 @@ import uk.ac.wellcome.messaging.worker.steps.MonitoringProcessor
 
 import scala.concurrent.Future
 
-class AlpakkaSQSWorker[
-  Work,
-  MonitoringContext,
-  Summary,
-  MonitoringProcessorImpl <: MonitoringProcessor[Work, MonitoringContext]](
+class AlpakkaSQSWorker[Work, MonitoringContext, Summary](
   config: AlpakkaSQSWorkerConfig
 )(
   val doWork: Work => Future[Result[Summary]]
 )(implicit
-  val monitoringProcessor: MonitoringProcessorImpl,
+  val monitoringProcessor: MonitoringProcessor[Work, MonitoringContext],
   val as: ActorSystem,
   val wd: Decoder[Work],
   sc: AmazonSQSAsync,
@@ -60,11 +56,7 @@ object AlpakkaSQSWorker {
     sc: AmazonSQSAsync,
     as: ActorSystem,
     wd: Decoder[Work]) =
-    new AlpakkaSQSWorker[
-      Work,
-      MonitoringContext,
-      Summary,
-      MonitoringProcessor[Work, MonitoringContext]](
+    new AlpakkaSQSWorker[Work, MonitoringContext, Summary](
       config
     )(process)
 }
diff --git a/messaging/src/test/scala/uk/ac/wellcome/messaging/fixtures/worker/AlpakkaSQSWorkerFixtures.scala b/messaging/src/test/scala/uk/ac/wellcome/messaging/fixtures/worker/AlpakkaSQSWorkerFixtures.scala
index 5a3b691..bc263f1 100644
--- a/messaging/src/test/scala/uk/ac/wellcome/messaging/fixtures/worker/AlpakkaSQSWorkerFixtures.scala
+++ b/messaging/src/test/scala/uk/ac/wellcome/messaging/fixtures/worker/AlpakkaSQSWorkerFixtures.scala
@@ -8,7 +8,6 @@ import uk.ac.wellcome.messaging.fixtures.SQS
 import uk.ac.wellcome.messaging.fixtures.SQS.Queue
 import uk.ac.wellcome.messaging.fixtures.monitoring.metrics.MetricsFixtures
 import uk.ac.wellcome.messaging.sqsworker.alpakka.{AlpakkaSQSWorker, AlpakkaSQSWorkerConfig}
-import uk.ac.wellcome.messaging.worker.monitoring.metrics.MetricsMonitoringProcessor
 import uk.ac.wellcome.monitoring.MetricsConfig
 
 import scala.concurrent.ExecutionContext
@@ -33,7 +32,7 @@ trait AlpakkaSQSWorkerFixtures
     queue: Queue,
     process: TestInnerProcess,
     namespace: String = Random.alphanumeric take 10 mkString
-  )(testWith: TestWith[(AlpakkaSQSWorker[MyWork, MyContext, MySummary, MetricsMonitoringProcessor[MyWork, FakeMetricsMonitoringClient]],
+  )(testWith: TestWith[(AlpakkaSQSWorker[MyWork, MyContext, MySummary],
                         AlpakkaSQSWorkerConfig,
                         FakeMetricsMonitoringClient,
                         CallCounter),
@@ -52,7 +51,7 @@ trait AlpakkaSQSWorkerFixtures
         createResult(process, callCounter)(ec)(o)
 
       val worker =
-        new AlpakkaSQSWorker[MyWork, MyContext, MySummary, MetricsMonitoringProcessor[MyWork, FakeMetricsMonitoringClient]](config)(testProcess)
+        new AlpakkaSQSWorker[MyWork, MyContext, MySummary](config)(testProcess)
 
       testWith((worker, config, monitoringClient, callCounter))
     }


```

**Review by warrd — APPROVED** [review]
Looking nice :tada:

**Pull request merged**